crypto/src/utils.rs

```diff
@@ -0,0 +1,104 @@
+#![allow(dead_code)]
+use super::{gen_keypair, hash, PublicKey, SecretKey};
+use openssl::symm::{decrypt, encrypt, Cipher};
+use pem::{encode, parse, Pem};
+#[cfg(not(windows))]
+use std::os::unix::fs::OpenOptionsExt;
+use std::{
+    fs::{File, OpenOptions},
+    io::Error,
+    io::Read,
+    io::Write,
+    path::Path,
+};
+const DEFAULT_KEY_PATH: &str = "../private_key_file.pem";
+const IV: Option<&[u8]> = Some(b"\x00\x01\x02\x03\x04\x05\x06\x07\x00\x01\x02\x03\x04\x05\x06\x07");
+
+#[derive(Copy, Clone)]
+enum KeyOperation {
+    ENCRYPT,
+    DECRYPT,
+}
+
+/// Creates a PEM file that contains encrypted `SecretKey` and returns `PublicKey` for the secret key.
+pub fn create_pem_file(path: Option<&str>, pass_phrase: &[u8]) -> Result<PublicKey, Error> {
+    let (pk, sk) = gen_keypair();
+    let sk_bytes = hex::decode(sk.to_hex()).expect("Couldn't decode secret key in bytes");
+    let enc_sk = key_processing(pass_phrase, KeyOperation::ENCRYPT, &sk_bytes)?;
+    let pem = Pem {
+        tag: "PRIVATE KEY".to_string(),
+        contents: enc_sk,
+    };
+
+    let mut pem_file = if cfg!(target_os = "windows") {
+        File::create(path.unwrap_or(DEFAULT_KEY_PATH))?
+    } else {
+        OpenOptions::new()
+            .create(true)
+            .write(true)
+            .mode(0o600)
+            .open(path.unwrap_or(DEFAULT_KEY_PATH))?
+    };
+    pem_file.write_all(encode(&pem).as_bytes())?;
+
+    Ok(pk)
+}
+
+/// Reads `SecretKey` from encrypted file located by path and returns it.
+pub fn read_key_from_pem<P: AsRef<Path>>(path: P, pass_phrase: &[u8]) -> Result<SecretKey, Error> {
+    let mut key_file = File::open(path)?;
```

Comment:
Should we check file permissions before opening? And panic when permission is not equal to 600?

Reply:
@pavel-mukhanov i suppose you're right. Good point 👍🏻

Reply:
Erm... Should we really? Why do we bother, if user wants to set something else instead of 600 file permission?

Reply:
It may mean that the key was compromised.

Reply:
Well, I can't advocate for security best practices.. If such behavior is common for this kind of case, well then. Just for me it looked a little bit strange, that we limit user workflow that way. What if an administrator decides that such files can be read (at least) by some trusted group?
The key should already be protected with a password.

Reply:
I also don't like this proposal. We can make a warning, but wrong permissions should not cause panic